File: go_core/internal/service/mission_approval_guard.go
package service

import (
	"errors"
	"fmt"
	"go_core/internal/domain"
)

// ApproveMission enforces that a MissionDraft can only be approved if in PENDING_APPROVAL state.
func ApproveMission(
	draft domain.MissionDraft,
	approvedBy string,
	approvedAtMs int64,
	planSnapshotID string,
) (domain.SearchMissionPlan, error) {
	if draft.ValidationStatus != "PENDING_APPROVAL" {
		return domain.SearchMissionPlan{}, fmt.Errorf("cannot approve draft in status '%s'; expected 'PENDING_APPROVAL'", draft.ValidationStatus)
	}

	if approvedBy == "" {
		return domain.SearchMissionPlan{}, errors.New("approver operator ID cannot be empty")
	}

	return domain.SearchMissionPlan{
		MissionID:      draft.MissionID,
		SearchArea:     draft.Request.SearchArea,
		SearchMethod:   draft.Request.SearchMethod,
		ApprovedBy:     approvedBy,
		ApprovedAtMs:   approvedAtMs,
		PlanSnapshotID: planSnapshotID,
	}, nil
}

File: src/service/mission_approval_guard.py
"""Mission approval guard.

Enforces that a MissionDraft cannot become a SearchMissionPlan
without Mission Commander approval.
"""

from __future__ import annotations

from domain_types.mission import MissionDraft, SearchMissionPlan


def approve_mission(
    draft: MissionDraft,
    approved_by: str,
    approved_at_ms: int,
    plan_snapshot_id: str,
) -> SearchMissionPlan:
    """Approve a mission draft and produce a frozen SearchMissionPlan.

    Args:
        draft: The validated MissionDraft to approve.
        approved_by: ID of the Mission Commander granting approval.
        approved_at_ms: Deterministic timestamp of approval.
        plan_snapshot_id: Deterministic snapshot ID.

    Raises:
        ValueError: If the draft is not in PENDING_APPROVAL status.
    """
    if draft.validation_status != "PENDING_APPROVAL":
        raise ValueError(
            f"Cannot approve draft in status '{draft.validation_status}'. "
            f"Expected 'PENDING_APPROVAL'."
        )

    return SearchMissionPlan(
        mission_id=draft.mission_id,
        search_area=draft.request.search_area,
        search_method=draft.request.search_method,
        approved_by=approved_by,
        approved_at_ms=approved_at_ms,
        plan_snapshot_id=plan_snapshot_id,
    )
